csharp/src/Apache.Arrow/Ipc/DictionaryMemo.cs

```diff
@@ -24,13 +24,13 @@ class DictionaryMemo
     {
         private readonly Dictionary<long, IArrowArray> _idToDictionary;
         private readonly Dictionary<long, IArrowType> _idToValueType;
-        private readonly Dictionary<Field, long> _fieldToId;
+        private readonly Dictionary<string, long> _fieldToId;
```

Comment:
I don't think this change is technically correct because it ignores the fact that e.g. two nested structs could have dictionary-encoded fields with the same name and these would now collide. Obviously the existing implementation isn't great either in that it implicitly relies on reference equality (which I assume is what causes problems for Flight). I don't have any immediate suggestions though.

Reply:
Thanks for this. When I have time to get back to this I will write a test for this and look into it.

Reply:
You can check out my attempt at an implementation at https://github.com/CurtHagenlocher/arrow/tree/dev/curth/FlightDictionaries. It changes the write side of flight server so that the same schema is used to collect the dictionaries for each batch. This continues to let reference equality work for calculating dictionary IDs. I also refactored the "Post" changes to be (what I think is) a little cleaner.

Disclaimer: I am not very familiar with the Flight code or protocol, but your changed tests are passing.